Ensuring aligned Straws are placed correctly by AlignedTrackerMaker (+ non-const Straw accessor)

ryuwd opened the pull request:
This should hopefully fix #77. All comments are welcome. An explanation of my changes:
- It seems like HepTransforms are being combined (+applied) in an unintended order. **adding in parentheses** in those expressions appears to correct the behaviour. My reasoning is:
    - A HepTransform is a GeneralUtilities object that describes a **rotation (about 0,0,0) followed by a translation** that may be applied to a CLHEP::Hep3Vector
    - C++ expressions are evaluated left-to-right. (further explanation below)
    - a straw position vector held by a panel shifted (using TrkAlignPanel rows) does not change by (dx,dy,dz) of the TrkAlignPanel row, but it does after I make the changes in this PR.
- Applying this logic to the alignment of straw positions, (and following the definition on https://mu2ewiki.fnal.gov/wiki/Alignment)
- (further explanation below)

I have run test jobs to see how the tracking responds to these changes. I applied the misalignments generated by JobConfig/reco/MT_MDC2018.C, which are plane misalignments proposed by Dave in Mu2e-doc-28722-v3. I will attach some quick plots of the de/dm chisq/ndof outputs and detsh branch information if anyone is interested.

ryuwd commented:
Here are some rudimentary plots showing the before and after misalignment comparisons of reduced-chisq and DOCA residuals:
- [chisqndof - TrkAnaNeg/trkana/dm](https://github.com/Mu2e/Offline/files/4137452/chisqndof.pdf)
- [chisqndof - TrkAnaNeg/trkana/de](https://github.com/Mu2e/Offline/files/4137453/chisqndof_DE.pdf)
- [DOCA residuals - TrkAnaNeg.detsh._resid](https://github.com/Mu2e/Offline/files/4137454/residuals.pdf)

kutschke commented:
Have you consulted with the original author about the change?

I don't know the intent of the original author but the following may be relevant.   The G4 convention is that volumes are placed by specifying passive transformations, not active transformations.  The relationship between the two is that each is the inverse of the other - which implies reversing the order of operations.  

At a minimum we should understand the intentions of the original author prior to making this change.

rlcee commented:
I feel strongly about Proditions, in particular, I want to maintain the thread-safe functions and I want all Proditions entities to have the same class structure and interface, such as "fromDb".  I don't feel strongly what is it called, just that 100% of Proditions entities are structured and interfaced the same way.  If they diverge, maintenance will be much harder, so I will insist on this point.  After that, I consider the contents of AlignedTrackerMaker to be the responsibility of the tracking group. I wrote it as an (unvalidated) example code.  If you ask me, I conceived of it as A*(B*(C*v)), and forgot the parens, but I don't see how reversing the order is a valid alternative unless there are canceling bugs.  It will take a me a few days to get back up to speed and make a detailed case, if you want me to.  But in general, I consider it your concepts, code, and wiki page now.  Sorry for the bugs and thanks for finding them.  I imagined this validation process would occur after there were some alignment plots available, such as per-panel residuals, so the details of the effect would be more clear.

I'm not sure why you added the HepTransform copy constructors, but this is also up to the tracker group or general coding experts.

I don't have a strong opinion of how to deal with modifying the const Straw in the Tracker.  Making AlignedTrackerMaker a friend of Tracker fits existing patterns.  I usually just do what Rob recommends in these messy cases.

I see your editor has changed bracket positions.  I think your version is less common in our code, but Rob and Dave should comment.  We have a code standards page, but it needs review and updating.

ryuwd commented:
Hi Ray,

Thanks for the reply and sorry for not contacting you sooner (I wanted to make sure I had some changes to present first.)

I will go ahead and do some cleanup on these changes. I think you're right that the code layout, copy ctor and copy assignment changes are un-necessary.

I've thought things out a bit further and I'm also currently dissatisfied with my changes which reverse the order of operations and aim to change this soon, as it looks unintuitive and potentially wrong. I think I may well actually be performing a 'passive' alignment here unintentially (that Rob mentioned), which is why I see a 'sensible' response in chi-squared. So, you are right about this and parentheses (thanks for this!)

I've confirmed that Hep Transform is doing the correct thing:

https://github.com/Mu2e/Offline/blob/959b19c0bc86b08e74fc9bdee4b4afe6c7cd6030/GeneralUtilities/src/HepTransform.cc#L22-L26

If we treat A and B as both functions A(v), B(v),
A(v) = R_A * v + r_A
B(v) = R_B * v + r_B
with R_A, R_B as rotation matrices and r_A, r_B as translation vectors. If we want to transform v by B first, then A (i.e. ABv), we write
A(B(v)) = R_A * ( R_B * v + r_B ) + r_A

which expands to
A(B(v)) = (R_A * R_B) * v + (R_A * r_B + r_A)

so HepTransform is correct in combining A with B such that B comes before A, **therefore the true fix is not just to reverse the operands as I have done, apologies.**

#### But in the current HEAD we effectively have (due to left-to-right associativity)

(((A * B) * C) * D) ...
`((((align_tracker * plane_to_tracker) * align_plane) * panel_to_plane) * align_panel)`
which is different to
`(A * (B * (C * D)))`

so e.g. align_tracker has a RHS of plane_to_tracker ONLY (which is wrong) when the RHS should be the combination of `plane_to_tracker * (align_plane * (panel_to_plane * align_panel))` which would be correct.

So what I did was reverse the transforms, which produced this effect:
`((((align_panel * panel_to_plane) * align_plane) * plane_to_tracker) * align_tracker)`
which is not correct for an active alignment. (but might actually be correct for passive alignment, which is not what we're after.)

One potential idea would be to have HepTransforms act on other HepTransforms or Hep3Vector via a `function()` overload, which makes the intention and nesting effect clearer and removes the confusion caused by left to right associativity.

I will make the cleanup and add the parentheses now. Thanks for the help and apologies for the confusion caused so far

Thanks, Ryun

ryuwd commented:
Changes made - the diffs are much shorter and easier to read as well. Here are the most recent plots (same misalignments as before):

## Current master (without the changes in this PR):
![chisqndof_de](https://user-images.githubusercontent.com/56410978/73751109-4c3c4800-4756-11ea-90ed-1aa47c30b919.png)
![residuals](https://user-images.githubusercontent.com/56410978/73750980-11d2ab00-4756-11ea-93c8-cfec6f7a89a4.png)

## After changes:
![chisqndof_de](https://user-images.githubusercontent.com/56410978/73751092-45153a00-4756-11ea-87ee-bd5d9e11dfe0.png)
![residuals](https://user-images.githubusercontent.com/56410978/73751114-53635600-4756-11ea-9759-20e5f01a4461.png)

kutschke commented:
Hi Ryun,

  A quick question: if we merge in your pull request, is it expected to change any default behaviours?

Rob


> On Feb 3, 2020, at 11:50 PM, ryuwd <notifications@github.com> wrote:
> 
> Hi Ray,
> 
> Thanks for the reply and sorry for not contacting you sooner (I wanted to make sure I had some changes to present first.)
> 
> I will go ahead and do some cleanup on these changes. I think you're right that the code layout, copy ctor and copy assignment changes are un-necessary.
> 
> I've thought things out a bit further and I'm also currently dissatisfied with my changes which reverse the order of operations and aim to change this soon, as it looks unintuitive and potentially wrong. I think I may well actually be performing a 'passive' alignment here unintentially (that Rob mentioned), which is why I see a 'sensible' response in chi-squared. So, you are right about this and parentheses (thanks for this!)
> 
> I've confirmed that Hep Transform is doing the correct thing:
> 
> https://github.com/Mu2e/Offline/blob/959b19c0bc86b08e74fc9bdee4b4afe6c7cd6030/GeneralUtilities/src/HepTransform.cc#L22-L26
> 
> If we treat A and B as both functions A(v), B(v),
> A(v) = R_A * v + r_A
> B(v) = R_B * v + r_B
> with R_A, R_B as rotation matrices and r_A, r_B as translation vectors. If we want to transform v by B first, then A (i.e. ABv), we write
> A(B(v)) = R_A * ( R_B * v + r_B ) + r_A
> 
> which expands to
> A(B(v)) = (R_A * R_B) * v + (R_A * r_B + r_A)
> 
> so HepTransform is correct in combining A with B such that B comes before A, therefore the true fix is not just to reverse the operands as I have done, apologies.
> 
> But in the current HEAD we effectively have (due to left-to-right associativity)
> 
> (((A * B) * C) * D) ...
> ((((align_tracker * plane_to_tracker) * align_plane) * panel_to_plane) * align_panel)
> which is different to
> (A * (B * (C * D)))
> 
> so e.g. align_tracker has a RHS of plane_to_tracker ONLY (which is wrong) when the RHS should be the combination of plane_to_tracker * (align_plane * (panel_to_plane * align_panel)) which would be correct.
> 
> So what I did was reverse the transforms, which produced this effect:
> ((((align_panel * panel_to_plane) * align_plane) * plane_to_tracker) * align_tracker)
> which is not correct for an active alignment. (but might actually be correct for passive alignment, which is not what we're after.)
> 
> One potential idea would be to have HepTransforms act on other HepTransforms or Hep3Vector via a function() overload, which makes the intention and nesting effect clearer and removes the confusion caused by left to right associativity.
> 
> I will make the cleanup and add the parentheses now. Thanks for the help and apologies for the confusion caused so far
> 
> Thanks, Ryun
> 
> —
> You are receiving this because your review was requested.
> Reply to this email directly, view it on GitHub, or unsubscribe.
>

brownd1978 commented:
Maybe I can answer.  We don't apply misalignments in any default process
currently, so no change in Validation etc. is expected.

On Tue, Feb 4, 2020 at 8:38 AM Rob Kutschke <notifications@github.com>
wrote:

> Hi Ryun,
>
> A quick question: if we merge in your pull request, is it expected to
> change any default behaviours?
>
> Rob
>
>
> > On Feb 3, 2020, at 11:50 PM, ryuwd <notifications@github.com> wrote:
> >
> > Hi Ray,
> >
> > Thanks for the reply and sorry for not contacting you sooner (I wanted
> to make sure I had some changes to present first.)
> >
> > I will go ahead and do some cleanup on these changes. I think you're
> right that the code layout, copy ctor and copy assignment changes are
> un-necessary.
> >
> > I've thought things out a bit further and I'm also currently
> dissatisfied with my changes which reverse the order of operations and aim
> to change this soon, as it looks unintuitive and potentially wrong. I think
> I may well actually be performing a 'passive' alignment here unintentially
> (that Rob mentioned), which is why I see a 'sensible' response in
> chi-squared. So, you are right about this and parentheses (thanks for
> this!)
> >
> > I've confirmed that Hep Transform is doing the correct thing:
> >
> >
> https://github.com/Mu2e/Offline/blob/959b19c0bc86b08e74fc9bdee4b4afe6c7cd6030/GeneralUtilities/src/HepTransform.cc#L22-L26
> >
> > If we treat A and B as both functions A(v), B(v),
> > A(v) = R_A * v + r_A
> > B(v) = R_B * v + r_B
> > with R_A, R_B as rotation matrices and r_A, r_B as translation vectors.
> If we want to transform v by B first, then A (i.e. ABv), we write
> > A(B(v)) = R_A * ( R_B * v + r_B ) + r_A
> >
> > which expands to
> > A(B(v)) = (R_A * R_B) * v + (R_A * r_B + r_A)
> >
> > so HepTransform is correct in combining A with B such that B comes
> before A, therefore the true fix is not just to reverse the operands as I
> have done, apologies.
> >
> > But in the current HEAD we effectively have (due to left-to-right
> associativity)
> >
> > (((A * B) * C) * D) ...
> > ((((align_tracker * plane_to_tracker) * align_plane) * panel_to_plane) *
> align_panel)
> > which is different to
> > (A * (B * (C * D)))
> >
> > so e.g. align_tracker has a RHS of plane_to_tracker ONLY (which is
> wrong) when the RHS should be the combination of plane_to_tracker *
> (align_plane * (panel_to_plane * align_panel)) which would be correct.
> >
> > So what I did was reverse the transforms, which produced this effect:
> > ((((align_panel * panel_to_plane) * align_plane) * plane_to_tracker) *
> align_tracker)
> > which is not correct for an active alignment. (but might actually be
> correct for passive alignment, which is not what we're after.)
> >
> > One potential idea would be to have HepTransforms act on other
> HepTransforms or Hep3Vector via a function() overload, which makes the
> intention and nesting effect clearer and removes the confusion caused by
> left to right associativity.
> >
> > I will make the cleanup and add the parentheses now. Thanks for the help
> and apologies for the confusion caused so far
> >
> > Thanks, Ryun
> >
> > —
> > You are receiving this because your review was requested.
> > Reply to this email directly, view it on GitHub, or unsubscribe.
> >
>
> —
> You are receiving this because you were assigned.
> Reply to this email directly, view it on GitHub
> <https://github.com/Mu2e/Offline/pull/131?email_source=notifications&email_token=ABAH5773MMXEZ4BHWJ54GWTRBGKM5A5CNFSM4KOAY222YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOEKYJHII#issuecomment-581997473>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ABAH572CXB4I5QF3Y7LXEYLRBGKM5ANCNFSM4KOAY22Q>
> .
>


-- 
David Nathan Brown            Dave_Brown@lbl.gov
Office Phone (510) 486-7261  Fax  495-2957
Lawrence Berkeley National Lab
MS 50R5008 (50-6026C)  Berkeley, CA 94720

kutschke commented:
HI Ryun,

  Thanks for responding to Ray's comments yesterday.   I need to find time to think carefully about order of operations issues.
But I have a few additional comments on simpler issues.

  About the non-const straw accessor,  I like the way you implemented the non-const straw accessor by reusing the const accessor.  A few small points:

  1) Please get rid of an unnecessary cast.  This will work
           return const_cast<Straw &>(this->getStraw(strawId));
      It's legal to call const member function of a non-const object.

  2) I don't think that it needs the comments you added.

  3) It does not need to be private; the only code that ever sees a non-const Tracker
      is the two Maker classes.  I suggest placing it right after the matching const function.

  4) If it were not for the indirection, I would recommend you just modify the data member directly.
      But because of the indirection I think you made the right choice.

   About copy constructors: if the compiler written copy constructor would be correct, please do not write one yourself; read about the C++ rule of 5.

    Thanks for backing out the whitespace changes.  Our request is that when you work on other people's code to follow their code layout practices.  If you are taking over code from someone else, we recommend that you make one commit that is formatting only.  Then make real changes in subsequent commits.

  We have discussed using an automated code formatter or code format checker as part of the commit or pull-request process. We have not had the time to make the investment in that - the big issue is converging on an agreed upon standard format.

     Rob


> On Feb 3, 2020, at 11:50 PM, ryuwd <notifications@github.com> wrote:
> 
> Hi Ray,
> 
> Thanks for the reply and sorry for not contacting you sooner (I wanted to make sure I had some changes to present first.)
> 
> I will go ahead and do some cleanup on these changes. I think you're right that the code layout, copy ctor and copy assignment changes are un-necessary.
> 
> I've thought things out a bit further and I'm also currently dissatisfied with my changes which reverse the order of operations and aim to change this soon, as it looks unintuitive and potentially wrong. I think I may well actually be performing a 'passive' alignment here unintentially (that Rob mentioned), which is why I see a 'sensible' response in chi-squared. So, you are right about this and parentheses (thanks for this!)
> 
> I've confirmed that Hep Transform is doing the correct thing:
> 
> https://github.com/Mu2e/Offline/blob/959b19c0bc86b08e74fc9bdee4b4afe6c7cd6030/GeneralUtilities/src/HepTransform.cc#L22-L26
> 
> If we treat A and B as both functions A(v), B(v),
> A(v) = R_A * v + r_A
> B(v) = R_B * v + r_B
> with R_A, R_B as rotation matrices and r_A, r_B as translation vectors. If we want to transform v by B first, then A (i.e. ABv), we write
> A(B(v)) = R_A * ( R_B * v + r_B ) + r_A
> 
> which expands to
> A(B(v)) = (R_A * R_B) * v + (R_A * r_B + r_A)
> 
> so HepTransform is correct in combining A with B such that B comes before A, therefore the true fix is not just to reverse the operands as I have done, apologies.
> 
> But in the current HEAD we effectively have (due to left-to-right associativity)
> 
> (((A * B) * C) * D) ...
> ((((align_tracker * plane_to_tracker) * align_plane) * panel_to_plane) * align_panel)
> which is different to
> (A * (B * (C * D)))
> 
> so e.g. align_tracker has a RHS of plane_to_tracker ONLY (which is wrong) when the RHS should be the combination of plane_to_tracker * (align_plane * (panel_to_plane * align_panel)) which would be correct.
> 
> So what I did was reverse the transforms, which produced this effect:
> ((((align_panel * panel_to_plane) * align_plane) * plane_to_tracker) * align_tracker)
> which is not correct for an active alignment. (but might actually be correct for passive alignment, which is not what we're after.)
> 
> One potential idea would be to have HepTransforms act on other HepTransforms or Hep3Vector via a function() overload, which makes the intention and nesting effect clearer and removes the confusion caused by left to right associativity.
> 
> I will make the cleanup and add the parentheses now. Thanks for the help and apologies for the confusion caused so far
> 
> Thanks, Ryun
> 
> —
> You are receiving this because your review was requested.
> Reply to this email directly, view it on GitHub, or unsubscribe.
>

ryuwd commented:
Hi Rob, 

> Maybe I can answer.  We don't apply misalignments in any default process
currently, so no change in Validation etc. is expected.

I agree with Dave here. 

Thanks for your comments so far,

> 
> HI Ryun, Thanks for responding to Ray's comments yesterday. I need to find time to think carefully about order of operations issues. But I have a few additional comments on simpler issues. About the non-const straw accessor, I like the way you implemented the non-const straw accessor by reusing the const accessor. 
> A few small points: 1) Please get rid of an unnecessary cast. This will work return const_cast<Straw &>(this->getStraw(strawId)); It's legal to call const member function of a non-const object. 
> 2) I don't think that it needs the comments you added. 
> 3) It does not need to be private; the only code that ever sees a non-const Tracker is the two Maker classes. I suggest placing it right after the matching const function. 
> 4) If it were not for the indirection, I would recommend you just modify the data member directly. But because of the indirection I think you made the right choice. 
> About copy constructors: if the compiler written copy constructor would be correct, please do not write one yourself; read about the C++ rule of 5. 

I'm glad this solution is acceptable - an alternative may have involved some fishy pointer arithmetic and other potentially repeated code. I had also added the static_cast to be cautious - but see now this was un-necessary. I've also reverted those copy constructor + whitespace changes too. Thanks!

> Thanks for backing out the whitespace changes. Our request is that when you work on other people's code to follow their code layout practices. If you are taking over code from someone else, we recommend that you make one commit that is formatting only. Then make real changes in subsequent commits. We have discussed using an automated code formatter or code format checker as part of the commit or pull-request process. We have not had the time to make the investment in that - the big issue is converging on an agreed upon standard format. Rob

Great - I will make one more commit then, to move the non-const accessor, remove those comments and the static_cast. 

Thanks also for the code style advice again - Ray mentioned there were code standards written up and available? Do you know where I could read these?

rlcee commented:
> Thanks also for the code style advice again - Ray mentioned there were code standards written up and available? Do you know where I could read these?

https://mu2ewiki.fnal.gov/wiki/Code#Standards_and_Practices

ryuwd commented:
My latest commit incorporates Rob's suggestions.

kutschke commented on TrackerConditions/src/AlignedTrackerMaker.cc:
```diff
@@ -80,13 +80,12 @@ namespace mu2e {
 
 	for(size_t istr=0; istr< StrawId::_nstraws; istr++) {
-	  // need to const cast so we can update geometry - fixme
-	  Straw& straw = const_cast<Straw &>(panel.getStraw(istr));
```
I will approve this now.  In a future iteration I recommend adding the non-const accessor to the Panel class.  As before only the two Maker classes ever see a non-const tracker or a non-const panel.
